cpp/src/strings/copying/copying.cu

```diff
@@ -0,0 +1,245 @@
+ * limitations under the License.
+ */
+
+#include <bitmask/valid_if.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/copying.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include "../utilities.hpp"
+#include "../utilities.cuh"
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/for_each.h>
+#include <thrust/sequence.h>
+#include <thrust/scatter.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+// new strings column from subset of this strings instance
+std::unique_ptr<cudf::column> slice( strings_column_view strings,
+                                     size_type start, size_type end,
+                                     size_type step, cudaStream_t stream,
+                                     rmm::mr::device_memory_resource* mr  )
+{
+    size_type num_strings = strings.size();
+    CUDF_EXPECTS( num_strings > 0, "Column has no strings.");
+    if( step == 0 )
+        step = 1;
+    CUDF_EXPECTS( step > 0, "Parameter step must be positive integer.");
+    if( end < 0 || end > num_strings )
+        end = num_strings;
+    CUDF_EXPECTS( ((start >= 0) && (start < end)), "Invalid start parameter value.");
+    num_strings = cudf::util::round_up_safe<size_type>((end - start),step);
+    //
+    auto execpol = rmm::exec_policy(stream);
+    // build indices
+    thrust::device_vector<size_type> indices(num_strings);
+    thrust::sequence( execpol->on(stream), indices.begin(), indices.end(), start, step );
+    // create a column_view as a wrapper of these indices
+    column_view indices_view( data_type{INT32}, num_strings, indices.data().get(), nullptr, 0 );
+    // build a new strings column from the indices
+    return gather(strings, indices_view, stream, mr);
+}
+
+// return new strings column with strings from this instance as specified by the indices
+std::unique_ptr<cudf::column> gather( strings_column_view strings,
+                                      column_view gather_map, cudaStream_t stream,
+                                      rmm::mr::device_memory_resource* mr  )
+{
+    CUDF_EXPECTS( strings.size() > 0, "Column has no strings.");
+    auto num_strings = gather_map.size();
+    // TODO use index-normalizing iterator to allow any numeric type for gather_map
+    CUDF_EXPECTS( gather_map.type().id()==cudf::INT32, "strings gather method only supports int32 indices right now");
+    auto d_indices = gather_map.data<int32_t>();
+
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_column = *strings_column;
+    auto d_offsets = strings.offsets().data<int32_t>();
+
+    // build offsets column
+    auto offsets_transformer = [d_column, d_offsets] __device__ (size_type idx) {
+            if( d_column.nullable() && d_column.is_null(idx) )
```

Comment:
I changed `is_null()` in https://github.com/rapidsai/cudf/pull/3008 to first check for `nullable()` so that we don't have to do this all over the place. To get the old behavior of `is_null()` that doesn't check for `nullable()`, you can use `is_null_nocheck()`. 
```suggestion
            if(d_column.is_null(idx) )
```